src/scipp/core/bin_remapping.py

```diff
@@ -1,114 +1,124 @@
+
+
+def _combine_bins(var: Variable, coords: Dict[str, Variable], edges: List[Variable],
+                  groups: List[Variable], dim: Dims) -> Dict[str, Variable]:
+    from .binning import make_binned
+    # Overview
+    # --------
+    # The purpose of this code is to combine existing bins, but in a more general
+    # manner than `concat`, which combines all bins along a dimension. Here we operate
+    # more like `groupby`, which combines selected subsets and creates a new output dim.
+    #
+    # Approach
+    # --------
+    # The algorithm works conceptually similar to `_concat_bins`, but with an additional
+    # step, calling `make_binned` for grouping within the erased dims. For the final
+    # output binning, instead of summing the input bin sizes over all erased dims, we
+    # sum only within the groups created by `make_binned`.
+
+    # Preserve subspace dim order of input data, instead of the one given by `dim`
+    changed_dims = [d for d in var.dims if d in concrete_dims(var, dim)]
+    unchanged_dims = [d for d in var.dims if d not in changed_dims]
+    changed_shape = [var.sizes[d] for d in changed_dims]
+    unchanged_shape = [var.sizes[d] for d in unchanged_dims]
+    changed_volume = prod(changed_shape)
+
+    # Move modified dims to innermost to ensure data is written in contiguous memory.
+    var = var.transpose(unchanged_dims + changed_dims)
+    params = DataArray(var.bins.size(), coords=coords)
+    params.attrs['begin'] = var.bins.constituents['begin'].copy()
+    params.attrs['end'] = var.bins.constituents['end'].copy()
+
+    # Sizes and begin/end indices of changed subspace
+    sub_sizes = index(changed_volume).broadcast(dims=unchanged_dims,
+                                                shape=unchanged_shape)
+    params = params.flatten(to=uuid.uuid4().hex)
+    params = _with_bin_sizes(params, sub_sizes)
+    # Apply desired binning/grouping to sizes and begin/end
+    params = make_binned(params, edges=edges, groups=groups)
+
+    # Setup view of source content with desired target bin order
+    source = _cpp._bins_no_validate(
+        data=var.bins.constituents['data'],
         dim=var.bins.constituents['dim'],
-        begin=out_begin,
-        end=out_end,
-    )
-
-    # Copy all bin contents, performing the actual reordering with the content buffer.
-    out[...] = var
-
-    # Setup output indices. This will have the "merged" bins, referencing the new
-    # contigous layout in the content buffer.
-    out_sizes = sizes.sum(dim)
-    out_end = cumsum(out_sizes)
-    out_begin = out_end - out_sizes
-    return _cpp._bins_no_validate(
-        data=out.bins.constituents['data'],
-        dim=out.bins.constituents['dim'],
-        begin=out_begin,
-        end=out_end,
+        begin=params.bins.constituents['data'].attrs['begin'],
+        end=params.bins.constituents['data'].attrs['end'],
     )
+    # Call `copy()` to reorder data and `_with_bin_sizes` to put in place new indices.
```

Comment:
Not sure I understand how `copy` re-orders data?

Reply:
We have reordered the begin/end indices and created `source` (in the statement before this). The reordering occurs since copying binned data will always make a contiguous mapping of bins (instead of keeping the order in the content buffer). I will add a more detailed comment.